File: doi.go
package main

import (
	"regexp"
	"strings"
)

var doiRegex = regexp.MustCompile(`(?i)^10\.\d{4,}(?:\.\d+)?/\S+$`)

// normalizeDOI strips known doi.org URL prefixes (case-insensitive) and
// lowercases the result. DOIs are officially case-insensitive, so storage
// and comparison use the lowercase form.
func normalizeDOI(input string) string {
	input = strings.TrimSpace(input)
	lower := strings.ToLower(input)
	for _, p := range []string{
		"https://doi.org/", "http://doi.org/",
		"https://dx.doi.org/", "http://dx.doi.org/",
		"dx.doi.org/", "doi.org/",
	} {
		if strings.HasPrefix(lower, p) {
			return lower[len(p):]
		}
	}
	return lower
}

File: doi.py
import re

DOI_REGEX = re.compile(r"^10\.\d{4,}(?:\.\d+)?/\S+$", re.IGNORECASE)

_URL_PREFIXES = (
    "https://doi.org/",
    "http://doi.org/",
    "https://dx.doi.org/",
    "http://dx.doi.org/",
    "dx.doi.org/",
    "doi.org/",
)


def normalize_doi(s: str) -> str:
    """Trim, lowercase, and strip known doi.org URL prefixes.

    DOIs are officially case-insensitive, so storage and comparison
    use the lowercase form.
    """
    s = s.strip().lower()
    for prefix in _URL_PREFIXES:
        if s.startswith(prefix):
            return s[len(prefix):]
    return s
